Cli gdal proximity (#12350)
OSGeo/gdal: doc/source/programs/gdal_raster_proximity.rst

.. _gdal_raster_proximity:

================================================================================
``gdal raster proximity``
================================================================================

.. versionadded:: 3.11

.. only:: html

    Produces a raster proximity map.

.. Index:: gdal raster proximity

Synopsis
--------

.. program-output:: gdal raster proximity --help-doc

Description
-----------

:program:`gdal raster proximity` generates a raster proximity map indicating the distance from
the center of each pixel to the center of the nearest pixel identified as a target pixel.
Target pixels are those in the source raster for which the raster pixel value is in the set of
target pixel values.

Standard options
++++++++++++++++

.. include:: gdal_options/of_raster_create_copy.rst

.. include:: gdal_options/co.rst
.. include:: gdal_options/overwrite.rst

.. include:: gdal_options/ot.rst

.. option:: -b, --band <BAND>

    Input band (1-based index)

> .. option:: --target-values <TARGET-VALUES>

Reviewer: What is the format of the list? Are ranges allowed?

Author: No, it would be a nice addition though.

Author: I have clarified the list format.

Reviewer 2: Using this in a pipeline after `gdal raster reclassify` would effectively allow ranges to be used as the target value.

    A list of target pixel values in the source image to be considered target pixels.
    If not specified, all non-zero pixels will be considered target pixels.

.. option:: --max-distance <MAX-DISTANCE>

    Maximum distance. The nodata value will be used for pixels beyond this distance (default: 0)

.. option:: --fixed-buffer <FIXED-BUFFER>

    Fixed buffer value (instead of the actual distance) (default: 0)

.. options:: --nodata <NODATA>

    Nodata value for the output raster. If not specified, the nodata value of the input band will be used.

.. options:: --respect-input-nodata

    When set, nodata pixels in the input raster will be treated as nodata in the output raster.

> 

Reviewer: Units are undocumented, right? `distanceUnits =  pixel|geo`.

Author: fixed

.. _gdal_raster_proximity:

================================================================================
``gdal raster proximity``
================================================================================

> .. versionadded:: 3.11

Reviewer 3: ```suggestion
.. versionadded:: 3.12
```

.. only:: html

    Produces a raster proximity map.

.. Index:: gdal raster proximity

Synopsis
--------

.. program-output:: gdal raster proximity --help-doc

Description
-----------

> :program:`gdal raster proximity` generates a raster proximity map indicating the distance from

Reviewer 2: ```suggestion
:program:`gdal raster proximity` generates a raster proximity map indicating the Cartesian distance from
```

the center of each pixel to the center of the nearest pixel identified as a target pixel.
Target pixels are those in the source raster for which the raster pixel value is in the set of
target pixel values.

Standard options
++++++++++++++++

.. include:: gdal_options/of_raster_create_copy.rst

.. include:: gdal_options/co.rst
.. include:: gdal_options/overwrite.rst

.. include:: gdal_options/ot.rst

.. option:: -b, --band <BAND>

    Input band (1-based index)

.. option:: --target-values <TARGET-VALUES>

    A single value or a comma separated list of target pixel values in the source image
    to be considered target pixels.
    If not specified, all non-zero pixels will be considered target pixels.

.. option:: --max-distance <MAX-DISTANCE>

>     Maximum distance. The nodata value will be used for pixels beyond this distance.

Reviewer 2: ```suggestion
    Maximum distance to search for a target pixel. The NoData value will be output if no target pixel is found within this distance.
```

    Distance is interpreted in pixels unless `--distance-units geo` is specified.

.. option:: --distance-units <pixel|geo>

    Units for the distance. If `geo` is specified, the distance will be calculated in
    georeferenced units (meters or degrees) using the pixel size of the input raster.
    Otherwise, the distance is interpreted in pixels.

> .. option:: --fixed-buffer <FIXED-BUFFER>

Reviewer 2: This option name isn't very clear to me (although the description is.) Would `--burn` be appropriate, as in `gdal vector rasterize` ? Or `--write-value` ?

Author: I'm happy to change it, `burn` means nothing to me (besides reminding of a fire). Wouldn't perhaps `--fixed-value` be more appropriate to convene the meaning that we are writing a fixed value instead of the varying distance?

Reviewer 2: `--fixed-value` sounds good.

>     Define a fixed value to be written in the output pixels that are within the
>     specified maximum distance from the target pixels, instead of the actual distance.

Reviewer 2: ```suggestion
    Define a fixed value to be written to output pixels that are within :option:`--max-distance`
    from the target pixels, instead of the actual distance.
```

>     specified maximum distance from the target pixels, instead of the actual distance.

Reviewer: Could it be then just `fixed-value`?

Author: > Could it be then just `fixed-value`?

I like it!

> .. option:: --nodata <NODATA>

Reviewer 2: Use `--dst-nodata` for consistency with other commands ?

Author: I'm all in for consistency but we have a mixed situation already. Maybe worth filing a ticket to address the nodata argument altogether (and also the `NoData` camel case in the docs)?

![immagine](https://github.com/user-attachments/assets/61e7b672-33aa-4938-b6b1-94924921d398)

Author: Ticketed as: https://github.com/OSGeo/gdal/issues/12364

>     Nodata value for the output raster. If not specified, the nodata value of the input band will be used.

Reviewer 2: What happens if the input band does not have a NoData value?

Author: According to https://gdal.org/en/stable/programs/gdal_proximity.html 65536 is used but let me write a few tests for that: I doubt that it will work with `Byte` output type.

Author: Tests added and documentation expanded.

> .. option:: --respect-input-nodata

Reviewer 2: `--propagate-nodata` ?

Author: I have removed the argument completely.

    When set, nodata pixels in the input raster will be treated as nodata in the output raster.

Advanced options
++++++++++++++++

.. include:: gdal_options/oo.rst

.. include:: gdal_options/if.rst

Examples
--------

.. example::

>    :title: Proximity map of the of a raster with max distance of 3 pixels

Reviewer 2: ```suggestion
   :title: Proximity map of a raster with max distance of 3 pixels
```
